src/commands/registerCommands.ts

```diff
@@ -179,24 +181,84 @@ async function grammarAssociationCommand(documentURI: Uri, languageClient: Langu
   const fileUri = await window.showOpenDialog(options);
   if (fileUri && fileUri[0]) {
     // The XSD, DTD has been selected, get the proper syntax for binding this grammar file in the XML document.
-    const identifier = TextDocumentIdentifier.create(documentURI.toString());
     const grammarURI = fileUri[0];
     try {
-      const result = await commands.executeCommand(ServerCommandConstants.ASSOCIATE_GRAMMAR_INSERT, identifier, grammarURI.toString(), bindingType);
-      // Insert the proper syntax for binding
-      const lspTextDocumentEdit = <TextDocumentEdit>result;
-      const workEdits = new WorkspaceEdit();
-      for (const edit of lspTextDocumentEdit.edits) {
-        workEdits.replace(documentURI, languageClient.protocol2CodeConverter.asRange(edit.range), edit.newText);
+      if (bindingType == 'fileAssociation') {
+        // Bind grammar using file association
+        await bindWithFileAssociation(documentURI, grammarURI);
+      } else {
+        // Bind grammar using standard binding
+        await bindWithStandard(documentURI, grammarURI, bindingType, languageClient);
       }
-      workspace.applyEdit(workEdits); // apply the edits
-
     } catch (error) {
       window.showErrorMessage('Error during grammar binding: ' + error.message);
     };
   }
 }
 
+/**
+ * Perform grammar binding using file association through settings.json
+ *
+ * @param documentURI the URI of the current XML document
+ * @param grammarURI the URI of the user selected grammar file
+ */
+async function bindWithFileAssociation(documentURI: Uri, grammarURI: Uri) {
+  const absoluteGrammarFilePath = grammarURI.toString();
+  const currentFilename = path.basename(documentURI.toString());
+  const currentWorkspaceUri = getAbsoluteWorkspaceUri();
+  const relativeGrammarFilePath = getRelativeUri(currentWorkspaceUri, absoluteGrammarFilePath);
+  const grammarFilePath = absoluteGrammarFilePath.includes(currentWorkspaceUri) ? relativeGrammarFilePath : absoluteGrammarFilePath;
```

Comment:
I think currentWorkspaceUri can be undefined are you sur that we can call includes with undefined?

Reply:
if currentWorkspaceUri is undefined, it would make grammarFilePath the absoluteFilePath in line 211 correct?

Reply:
Ah i understand your question now you are referring to absoluteGrammarFilePath.includes(currentWorkspaceUri). If currentWorkspaceUri is undefined, it will be false, and the grammar file will use the absolute path. To me this makes sense since this can happen when a workspace is not open (i.e. a single open editor), so no relative path can be formed and therefore the absolute path should always be used.